### DISCUSSION TITLE
Survey answers polling with Imagemagick.... please point me to the right path

### DISCUSSION THREAD

**Discussion by NcMichel** [ImageMagick/ImageMagick#2544] [Help]
Hello all,
thanks for help I already received.

we are part of a local community project and would like to collect polling results at least half-automatically.

The survey looks like this:
![Qvierge](https://user-images.githubusercontent.com/56865796/92722817-bb3be800-f39a-11ea-923a-4f5cbad276dd.jpg)

The answers are scanned and get back as pdf file. I convert pdf to those jpg with

_convert -density 160 *.pdf -threshold 70% -type bilevel -deskew 80 -trim  *.jpg_

![out-3](https://user-images.githubusercontent.com/56865796/92722898-d9a1e380-f39a-11ea-978b-c86d07bb16c1.jpg)
![bourail-14](https://user-images.githubusercontent.com/56865796/92722905-ddce0100-f39a-11ea-85f0-2c8eaeefdf9e.jpg)
![bourail-12](https://user-images.githubusercontent.com/56865796/92722910-deff2e00-f39a-11ea-9a27-2bc3f5cd8239.jpg)
![out-13](https://user-images.githubusercontent.com/56865796/92722912-e0305b00-f39a-11ea-8a57-1057c46e120b.jpg)
![out-12](https://user-images.githubusercontent.com/56865796/92722916-e0c8f180-f39a-11ea-8247-c2424f659c61.jpg)

The idea is: How to track Checked, crossed and /or surrounded parts of answers ? How to compile a text file with those answers  ?

Giving that
- the scale of every scanned answer is very similar
- I can adjust the scale of the survey accordingly.
BUT
the offset can be annoying (even with -trim) since the scan quality si "as is".

I can use Macosx catalina
_Version: ImageMagick 7.0.10-29 Q16 x86_64 2020-09-10 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1) 
Delegates (built-in): bzlib fftw freetype gslib heic jng jp2 jpeg lcms lqr ltdl lzma openexr png ps tiff webp xml zlib_

Or I can use Ubuntu 20.4 version
_Version: ImageMagick 7.0.10-29 Q16 x86_64 2020-09-10 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype heic jbig jng jp2 jpeg lcms lqr lzma openexr pangocairo png rsvg tiff webp wmf x xml zlib_

computing time is irrelevant (at the moment volonteers fill .csv files by hand........)

At the moment I already try
-substract survey from results : fails due to offset
_compare Qvierge.jpg out-20.jpg -compose src diff.png_  
![diff](https://user-images.githubusercontent.com/56865796/92724175-c42db900-f39c-11ea-9d9a-70699954356c.png)

- find sub-images either with  normcrosscorr or compare -subimage-search : not found any subimage....

Therefore Im a bit lost in that matter. If one of you could point me the right way, Ill be more than grateful.

Michel

**Comment by snibgo** [comment]
> The answers are scanned and get back as pdf file. I convert pdf to those jpg with
convert -density 160 *.pdf -threshold 70% -type bilevel -deskew 80 -trim *.jpg

This is a poor workflow. I suggest you avoid using PDF. If you must use it, then remember that each page in a PDF file is a wrapper around at least one raster image. If you read the PDF with IM, each page is re-rasterized, probably re-sampling the embedded raster image. Instead, I suggest you use "pdfimages" to extract the embedded raster images from the PDF files.

> The idea is: How to track Checked, crossed and /or surrounded parts of answers ? How to compile a text file with those answers ?

From a page with answers, and a blank survey page, a possible workflow is:

1. Align the two images.

2. Slightly dilate (ie enlarge) the black marks in the blank survey page. Make the corresponding pixels white in the answered page.

3. Now you have an answer page with just the answers. You need a list of coordinates of rectangles where there might be answers. Crop the answer page to each rectangle. Each crop is either blank, or has a mark (an answer). This tells you when a tick box such as "OUI" or "NON" is ticked, or an answer such as "Je n'ai pas envie" is circled. It also tells you where free-format text has been used, such as un Q11.

**Reply by NcMichel** [reply]
Thank you so much
I use pdfimages and it works perfectly
the remaining pb is 
1. Align the two images
It seems I "may" use the right line of the surveys
![image](https://user-images.githubusercontent.com/56865796/92741404-2bedff00-f3b1-11ea-928a-1edfd292d8a7.png)
![image](https://user-images.githubusercontent.com/56865796/92741479-3f00cf00-f3b1-11ea-9efb-f381efbce10f.png)
------
the right line of blank survey look like
![image](https://user-images.githubusercontent.com/56865796/92741641-68b9f600-f3b1-11ea-9db0-791ad03d49ae.png)

Seems its a good problem for hough deskew ? is is possible to obtain angle + offset from that ?

Thanks
Michel

**Reply by NcMichel** [reply]
Hello
found this discussion
https://www.imagemagick.org/discourse-server/viewtopic.php?t=28183

seems to be not so far than what I need.
do you think its possible to adapt ?
thanks
Michel

**Comment by snibgo** [comment]
What version of IM, on what platform? If you use IM v7, I suggest you use "magick", not "convert".

We can use "-morphology" to find corners.
```
magick rightEdge.png -morphology Thinning:-1 Corners -colorspace Gray -threshold 50% +write x0.png -define identify:locate=minimum -define identify:limit=1 info:

Channel minimum locations:
  Gray: 0 (0) 189,142
  Alpha: 0 (0) 189,142
```
So the first corner is at (189,142). Invert the image and repeat to find the bottom corner. Do the same on the blank survey page. Now you have two points on the answer sheet that need to be moved to two coordinates from the blank sheet (or vice versa).

Do the same for points on the left, and you have four points. This is enough for a "-distort SRT perspective" transformation.

That's how I would do it. But "-deskew" might help, or Hough lines.

**Reply by NcMichel** [reply]
Sorry to ask
Where can I find documentation concerning
magick.... -morphology

since 
magick 'college-006.png[300x300+130+60]' -morphology Thinning:-1 Corners -colorspace Gray -threshold 50% +write x0.png -define identify:locate=minimum -define identify:limit=1 info:

returns
  Gray: 0 (0) 36,153
  Alpha: 0 (0) 36,153

and 
magick 'college-006.png' -fill red -draw 'rectangle 36,153 46,163'  
returns following image....... doesnt looks like an edge....


![image](https://user-images.githubusercontent.com/56865796/92879520-b11ff500-f43f-11ea-9663-5e1a52f96bd7.png)

**Reply by fmw42** [reply]
See https://imagemagick.org/Usage/morphology/

**Reply by snibgo** [reply]
On morphology, see http://www.imagemagick.org/Usage/morphology/

x0.png is an intermediate result, to help you see what is happening. It has no other purpose. Did you look at it?

You crop the input with offsets, and find a corner within that crop. But you use those coordinates to draw a red mark in the original input. Can you see that this is wrong? You need to add the offsets from the crop.

**Reply by NcMichel** [reply]
Ok, thanks for the crop problem, sorry....
BTW x0.png is absolutely blank. with every image i did test.

**Reply by NcMichel** [reply]
Just to let you know how it works at the moment. I write a little script to plot the points.
Yellow area is the cropping, points are plotted as "red","blue","green","yellow","purple" in order of return.
Original survey:
![image](https://user-images.githubusercontent.com/56865796/92993356-86a76800-f523-11ea-9c05-2d69718f743c.png)

----- filled surveys (after deskew, but not better without) ----
![image](https://user-images.githubusercontent.com/56865796/92993368-9d4dbf00-f523-11ea-9d7d-c893cd208d79.png)
![college-013 R](https://user-images.githubusercontent.com/56865796/92993373-a5a5fa00-f523-11ea-93d5-37c94a07cf1b.png)
![bourail-021 R](https://user-images.githubusercontent.com/56865796/92993377-ab9bdb00-f523-11ea-8df3-c95c9e3a7a09.png)
![bourail-010 R](https://user-images.githubusercontent.com/56865796/92993412-e998ff00-f523-11ea-9e6c-d75133db473a.png)
![bourail-004 R](https://user-images.githubusercontent.com/56865796/92993418-f158a380-f523-11ea-8085-276887afc2d8.png)
![college-003 R](https://user-images.githubusercontent.com/56865796/92993422-fb7aa200-f523-11ea-989e-23442a33c671.png)

---- seems something very ugly happen with the -Morphology function ---
I have some results with "corners" Script from Fred
http://www.fmwconcepts.com/imagemagick/corners/index.php
unfortunately, this one find ways too much corners at the moment.

![college-023 C](https://user-images.githubusercontent.com/56865796/92993456-50b6b380-f524-11ea-978b-7b1ebc7517e5.png)
![college-012 C](https://user-images.githubusercontent.com/56865796/92993461-57452b00-f524-11ea-8085-e52a7a25995b.png)
![bourail-025 C](https://user-images.githubusercontent.com/56865796/92993463-5ad8b200-f524-11ea-9cbd-04076f5be2cf.png)
![bourail-016 C](https://user-images.githubusercontent.com/56865796/92993467-5f04cf80-f524-11ea-9eff-628621ec98c6.png)
---- Work still in progress ! ----

**Comment by NcMichel** [comment]
Version: ImageMagick 7.0.10-29 Q16 x86_64 2020-09-10 https://imagemagick.org
Thanks for your help.

**Comment by fmw42** [comment]
I suggest that you put fiducial marks in the corners of your form. Those can then be found (by corner detection or compare or correlation). Then warp one image to match the other via -distort perspective. Then erode (not dilate) since you have black lines and marks on white. Then do your differencing or whatever you want to extract your answers.  Alternately, use OpenCV with ORB or SIFT to locate matching points automatically and then do a warpPerspective and your differencing.

**Reply by NcMichel** [reply]
Hi
For next year survey we follow the idea to put fiducial marks.
is there a "best way" to detect those marks on images ?
I did some tests with stars 
![image](https://user-images.githubusercontent.com/56865796/95552689-b58bfd80-0a3f-11eb-984a-576e4d026e8f.png)
and this command
magick blank2_legacy.png  -colorspace RGB -fuzz 20%  -morphology close 'plus:15x15;cross'  crp.png
then finding objects in the resulting files.
Any other idea/way ?
Of course search for subimages seems to be irrelevant since the survey are distort by scanning process.
Thanks
Michel

**Comment by fmw42** [comment]
Correlation of your images will not whether from my script or compare because of rotation and scale differences.

Getting rid of the text and small features will likely help reduce the number of corners. You can try either morphology and/or connected components filtering or hough line transform to get the long straight lines and their intersections for your corners.

But you still need to be able to filter the corners to get corresponding ones for -distort rigidaffine so that you do not get skew warping only scale, rotation and translation.

I will be trying similar things tomorrow in Python/OpenCV.  

Fred

**Reply by NcMichel** [reply]
Hello
did you see this one 
https://www.imagemagick.org/discourse-server/viewtopic.php?t=28183

I try to adapt but the first left houghline seems always in bad position.
thanks
MJ

**Reply by fmw42** [reply]
URL not found error

**Reply by NcMichel** [reply]
seems somebody remove it
Here are the bash script used and the 2 sample images
They compare/resize/reoffset the 2 grids with houghlines methods and -compose 

BTW i find no source for this syntax
    **-format "X2B=%[fx:page.width-w-1]\nY2B=%[fx:page.height-h-1]\n" \**
is there any maual page ?

Thnaks everybody for your helo.
Michel

`#!/bin/bash

convert grid1.jpg -deskew 40 -crop 99%x99%+0+0 +repage +depth g1.png
convert grid2.jpg -deskew 40 -crop 99%x99%+0+0 +repage +depth g2.png


convert \
  g1.png \
  -negate \
  -background "#8ff" -fill red \
  -hough-lines 2x2+1000 \
  h1.png


convert \
  g2.png \
  -negate \
  -background "#8ff" -fill red \
  -hough-lines 2x2+1000 \
  h2.png


declare `convert \
  h1.png \
  \( -clone 0 \
    -fill Lime \
    -draw "color 0,0 floodfill" \
    -fill Black +opaque Lime \
    -bordercolor Black -trim \
    -format "X1A=%[fx:w+1]\nY1A=%[fx:h+1]\n" \
    +write info: \
    +delete \
  \) \
  \( -clone 0 \
    -flip -flop \
    -fill Lime \
    -draw "color 0,0 floodfill" \
    -fill Black +opaque Lime \
    -bordercolor Black -trim \
    -format "X1B=%[fx:page.width-w-1]\nY1B=%[fx:page.height-h-1]\n" \
    +write info: \
    +delete \
  \) \
  NULL:`

declare `convert \
  h2.png \
  \( -clone 0 \
    -fill Lime \
    -draw "color 0,0 floodfill" \
    -fill Black +opaque Lime \
    -bordercolor Black -trim \
    -format "X2A=%[fx:w+1]\nY2A=%[fx:h+1]\n" \
    +write info: \
    +delete \
  \) \
  \( -clone 0 \
    -flip -flop \
    -fill Lime \
    -draw "color 0,0 floodfill" \
    -fill Black +opaque Lime \
    -bordercolor Black -trim \
    -format "X2B=%[fx:page.width-w-1]\nY2B=%[fx:page.height-h-1]\n" \
    +write info: \
    +delete \
  \) \
  NULL:`

echo "$X2A,$Y2A,$X1A,$Y1A,$X2A,$Y2B,$X1A,$Y1B,$X2B,$Y2A,$X1B,$Y1A,$X2B,$Y2B,$X1B,$Y1B"

convert \
  g2.png \
  -distort Perspective \
$X2A,$Y2A,$X1A,$Y1A,\
$X2A,$Y2B,$X1A,$Y1B,\
$X2B,$Y2A,$X1B,$Y1A,\
$X2B,$Y2B,$X1B,$Y1B \
  g2d.png

convert \
  g1.png \
  g2d.png \
  -compose Darken -composite \
  gcomp.png`

![grid1](https://user-images.githubusercontent.com/56865796/93038082-9edbcc00-f676-11ea-9bee-ea621e55198c.jpg)
![grid2](https://user-images.githubusercontent.com/56865796/93038090-a4391680-f676-11ea-952f-598b98e19140.jpg)

**Reply by fmw42** [reply]
See bash declare at https://linuxhint.com/bash_declare_command/
